### ISSUE TITLE
Standalone: Audio Device revamping needed, OSX not having a device selection. device selection save/restore buggy

### ISSUE THREAD

**Issue by fab672000** [elieserdejesus/JamTaba#1144]
Hi guys, I have exciting changes that fix all of this now we have on OSX the device change supported but much more ...

For all platforms: we now get separate audio device input output selection!
and now no default channel for OSX portaudio wrapper code meaning that you can use any input channel on OSX too!

Super excited  toshare all this soon,

@Eli: I also have a patched portaudio lib on OSX that fixes the channel names tyext (that always loose last char), so that now if fully works as expected!

**Comment by elieserdejesus** [comment]
Hi @fab672000, thanks for your collaboration, I have 2 comments:

1 - Please create a Pull Request.

2 - *Now All plaforms get input and output separate device selection*. Maybe I'm wrong, but I think this is very problematic for windows users. In windows you have to choose ONE audio device (for input and output). We have no flexibility in windows like in OSX or Linux.  So, we will force windows users to work like in Mac or Linux, but the windows users database is bigger than others. 

![image](https://user-images.githubusercontent.com/1012741/48361768-5bcefd00-e689-11e8-82a6-ee58de141d7f.png)

[JamTaba Release stats](https://www.somsubhra.com/github-release-stats/?username=elieserdejesus&repository=JamTaba)

**Comment by fab672000** [comment]
Thanks Eli, I will create a PR  as soon as I finish testing and fix a few more things.

Something is telling me we will have more Mac OS X users once they get a device selection back (was disable in OSX and also channel was not selectable) :)

I may be wrong, but I don't think that would be a problem just letting users to have the choice to test different drivers if they would like to but it is still possible (and easy to change) to force the same driver for both input / output in Windows only, if we absolutely need it.
I'll test this with different ASIO drivers on PC and check this as well before I propose a PR, as you would expect :)

**Comment by elieserdejesus** [comment]
> I'll test this with different ASIO drivers on PC and check this as well before I propose a PR, as you would expect :)

Thanks @fab672000 !  Maybe we can have the *output device combobox* disabled (or hidden) in Windows only. So we can force the same I/O device programatically in Windows. 

I'm ok with some code like:

```
#ifdef OS_WIN
      outputComboBox.setVisible(false);
#endif 
```

**Comment by fab672000** [comment]
Yes Eli, we can do that if you want to. But the code is not the problem here. I was looking at what is done with other audio apps and vsti's wrappers and we can find both behaviors, ohmstudio and probably all apps built on the same (juce) framework use seperate input and output devices like I did.
But Kontakt seems to prefer your way of handling it.
I can see both advantages and drawbacks to both approaches:
On one hand it seems more robust / easy to use to limit the user interaction on Windows to only one device selection as for now (it may change in future), windows does not seem to cope very well with multiple device handling. On the other hand, this will always limit user to experiment other solutions that may have worked for her/him otherwise.
That said, I am not using as much windows audio than osx audio so I will agree as for today, but will make we can easily switch to dual devices whenever we see a clear advantage to do so in future.

How about linux ?
I have no real experience of device drivers there (was never very lucky with latency on these systems), should we not also force single devices ?

**Comment by fab672000** [comment]
Ok Eli I tested the unified driver mode under Windows I just implemented, a single const  definition is defaulting now to true in Windows only, making easy to experiment improve that platform in future while making sure windows users don't get unusable configurations today as you recommended.
Testing for regression ATM, once I'mm done I'll submit a PR tonight for you guys to kindly code review!

**Comment by elieserdejesus** [comment]
> Testing for regression ATM, once I'mm done I'll submit a PR tonight for you guys to kindly code review!

Nice @fab672000 , thanks!

**Comment by fab672000** [comment]
NP, working now on some hot stuff after progfiling the standalone code in windows : nice optimizations on sample buffers, and also made peaks computation much cheaper in terms of cpu use than it used to 
Stay tuned! :)

**Comment by elieserdejesus** [comment]
> nice optimizations on sample buffers, and also made peaks computation much cheaper in terms of cpu use than it used to
Stay tuned! I'd like to create some SSE2 / AVX2 FMA code with packed float calculation as a next step :)

I see the commit, very nice code tricks!!

I'm totally zero in SSE2 optimizations, I will learn with you. Thanks!!!!
